### ISSUE TITLE
The background color setting in the customizer does not work in some cases

### ISSUE THREAD

**Issue by ShariqKhan2012** [understrap/understrap#674]
## Issue is about..
  - [ ] a Bug Report

## Please tell us about your environment:

  - **WordPress Version:** 4.9.5
  - **OS Version:** Windows 7 
  - **Browser:** [ Chrome 66.0.3359 ]
  - **Language:** [ PHP | CSS]
  - **Node.js & NPM Versions:** Node.js [6.10.3] | NPM [5.3.0]

---

# Issue Overview for Bug reports
The background-color setting  (Customizer->Colors) does not seem to work if the default color (#fff) is selected and the _theme_vaiables.scss overrides $body-bg variable

## Steps to reproduce?
1. In _theme_variables.scss define:
`$body-bg: #F6FAF8;`

2. From customizer->Colors, select `#ffffff` as the background color.

3. Reload the page. The background does not change to #ffffff. Instead, the background color still takes the value defined in _theme_variable.scss i.e. #F6FAF8;

4. Now, from the customizer change the setting to some other color, for example, #ff0000.

5. Reload the page, and the background color surely changes to #ff000, as expected.

## *What* is affected by this bug?
Background color

## *Where* on the platform does it happen?
Customizer

## Expected behavior
Background color should be changed to whatever is set in the customizer.

### Other Comments
The issue does not occur if one does not override $bg-color. Neither does it occur when $bg-color is overridden but the color selected in customizer is set to anything other than the default (#000000)

**Comment by 0dp** [comment]
I am unable to reproduce the issue.

Please check the source in the browser.

You should see **WordPress** have included something like
```
<style type="text/css" id="custom-background-css">
body.custom-background { background-color: #ff0000; }
</style>
```
just before the closing head tag `</head>` as such the included inline script should override theme.min.css.

The theme actually does not handle the background color in the customizer, it's a built-in WP feature 😄

**Comment by ShariqKhan2012** [comment]
Thanks for checking it out. But the issue(?) is still reproducible. Perhaps I was not able to specify the steps clearly. I will give it a try again.

1. Override `$body-bg` in _theme_variables.scss. Set it to anything other than `#ffffff` 
`$body-bg: #F6FAF8;`
This step is very important. The issue is **not** reproducible without this step.
I will explain this later why this step is so important.

2. From customizer->Colors, select "default" as the background color and click 'Publish'.
The background changes to `#ffffff` as expected.

3. Now exit the customizer. The page will reload, but now the background is no more `#ffffff`. Instead it has reverted to `#F6FAF8` 

Now, let me explain why this happens.
In the file `/wp-includes/theme.php` the function `_custom_background_cb` has this:
```
	// $color is the saved custom color.
	// A default has to be specified in style.css. It will not be printed here.
	$color = get_background_color();

	if ( $color === get_theme_support( 'custom-background', 'default-color' ) ) {
		$color = false;
	}
```

So, if the color selected in the customizer matches the `default-color`, then the inline style will not be printed. `default-color` is the color set in the call to `add_theme_support( 'custom-background')`

Now, the default copy of understrap has `#ffffff` as the `default-color`.
```
		add_theme_support( 'custom-background', apply_filters( 'understrap_custom_background_args', array(
			'default-color' => 'ffffff',
			'default-image' => '',
		) ) );
```
So, now what happens here is:
1. Case 1: I override `$body-bg` to `#F6FAF8` in _theme_variables.scss, which gets reflected in _reboot.scss
```
body {
    background-color: #F6FAF8;
}
```

a) I go to Customizer and choose the dfeault color, then what happens is that the if clause (from the function `_custom_background_cb` in the file `/wp-includes/theme.php`) kicks in and the chosen color being equal to the default color (set in the add_theme_support), the inline style does not get printed.
In the absence of the inline style, body picks background color from the style-rule from _reboot.scss
```
body {
    background-color: #F6FAF8;
}
```
So, the default color does not get reflected.

b. On the other hand, if I choose some other color (eg #f00) in the customizer, then the if-clause does not kick in , because the chosen color (#f00) is not equal to the default color(#fff) and hence inline style gets printed. Thus the body element takes background color as #f00.

Case 2: I **DON'T** override `$body-bg` then in _reboot.scss,
```
body {
    background-color: #fff ;
}
```
a. I go to Customizer and choose the default color, then what happens is that the if clause (from the function `_custom_background_cb` in the file `/wp-includes/theme.php`) kicks in and the chosen color being equal to the default color (set in the add_theme_support), the inline style does not get printed.
In the absence of the inline style, body picks background color from the style-rule from _reboot.scss
```
body {
    background-color: #fff;
}
```
So, the default color **gets** reflected.

b. On the other hand, if I choose some other color (eg #f00) in the customizer, then the if-clause does not kick in , because the chosen color (#f00) is not equal to the default color(#fff) and hence inline style gets printed. Thus the body element takes background color as #f00.

After analysing the root cause behind this behavior, I understand this is because the background color needs to be specified in 2 places:
a) In our stylesheet
b) in `add_theme_support('background-color')`
and both of these should have the same color.

It was mistake on my part that I while I updated the background-color in` _theme_variables.scss`, I did not do the same in `add_theme_support('background-color')`

Hence, I come to the conclusion that this is not a bug per-se.
Perhaps, it could be added in the documentation?

Kindly feel free to close it as it is a non-issue.

**Comment by 0dp** [comment]
Thank you so much for your very thorough description. I am currently in the process of consuming large quantities of alcohol, So I will get back to when I'm not shit faced. ❤️

**Comment by 0dp** [comment]
Why am I on GitHub now???

**Comment by Thomas-A-Reinert** [comment]
This is a high-potential for becoming on of the best answers on github _ever_ :)

Anyway: @ShariqKhan2012  this is becoming a _bit_ too lengthy for me to read through everything, still it´s one of the most detailed descriptions, we ever got and we all really appreciate that! But i really just browsed over it and came to the bottom line that read _Kindly feel free to close it as it is a non-issue._

So is this **really** a show-stopper (bug on our side) that may conflict with every other user or can we neglect putting real effort into fixing it? From what I read, noone will ever meet that flaw. Either they will change background-color via SCSS or via customizer. From what i read this will be solved in 99% of all cases. Correct me if i´m wrong.

What you´re describing seems to be some kinda edge-case.

And now i´ll try to catch up to @0dp ´s condition :beers:

**Comment by ShariqKhan2012** [comment]
> And now i´ll try to catch up with @0dp ´s condition
Ha ha ha, best wishes. You guys are making the best of the weekend, it seems !!

> What you´re describing seems to be some kinda edge-case.
Basically, when someone (read theme developer) needs to change the default background color in his/her theme, they need to do it in 2 places: 
1. overriding the bg-color in one of their sass files
2. making a similar change in the call to `add_theme_support('background-color')` in `setup.php`.
making the change in only one of these places will lead to the issue when the user goes to the customizer and change the background color to default color.

As it is obvious now, this is not an understrap issue. Rather it was me failing to make the relevant changes in both places. Before I had investigated this, I thought it is understrap, so I created the issue, 
I stand corrected and say that this issue should be closed. Though I think, perhaps it could be added to the documentation. You guys are the pro here, so any decision you folks take will be wiser.

**TLDR:** A noobish developer using understrap wishes to change the default background color. Misses on of the changes required to do so, and thinks it is understrap issue. Jumps the gun and creates the issue before investigating it. Investigates it later and finds the issue holds no water. 
Writes a TLDR for 2 folks who are enjoying their weekend :)

**Comment by axlright** [comment]
Best ticket ever 💯 🥇 👍 🍺

**Comment by 0dp** [comment]
I still don't understand why I was up in da club checking out GitHub 🤓.

@thomas, hope you cought up and feels just as wrecked as I do 🙈

**Comment by Thomas-A-Reinert** [comment]
Lol, seems that party was going all to well at that time at least or the gals were like.. better check github before I get blind ;)

Yeap, cought up well but actually felt pretty good next day ;)
Had an audit with myself and try to improve today :P

@ShariqKhan2012 - I´ll close this issue for now but mark it for the upcoming documentation. Hope your explanation will find its way in there.
